pkg/internal/candidatebuffer/candidatebuffer.go

```diff
@@ -49,39 +49,26 @@ func (c *CandidateBuffer) emit() {
 	}
 }
 
+type OnNextCandidate func(types.RetrievalCandidate)
+
 // BufferStream consumes a stream of individual candidate results. When a new result comes in, a collection is started, and further results
 // are added to the collection until the specified bufferingTime has passed, at which point the collection is passed to the callback setup
 // when the Buffer was setup. The timer is reset and the collection emptied until another result comes in. This has the effect of grouping
 // results that occur in the same general time frame.
-func (c *CandidateBuffer) BufferStream(ctx context.Context, incoming <-chan types.FindCandidatesResult, bufferingTime time.Duration) error {
+func (c *CandidateBuffer) BufferStream(ctx context.Context, queueCandidates func(context.Context, OnNextCandidate) error, bufferingTime time.Duration) error {
 	var wg sync.WaitGroup
 	defer func() {
 		wg.Wait()
 	}()
-	for {
-		select {
-		case <-ctx.Done():
-			c.clear()
-			return ctx.Err()
-		case candidateResult, ok := <-incoming:
-			if !ok {
-				c.emit()
-				return nil
-			}
-			if candidateResult.Err != nil {
-				c.emit()
-				return candidateResult.Err
-			}
+	errChan := make(chan error, 1)
+	go func() {
```

Comment:
ah, so we've exchanged the async candidate finder goroutine for one in here so that the tests can be simpler

Reply:
yep!

But honestly: that's implemented in like 6 places, where as now it lives in one, with less lines of code even where it's added.

The longer I program in go the more I think: "defer your async till you need it, never return a channel when a callback will do"

Reply:
so yes, I started motivated by the tests, but I only went through with this change cause it felt like improvement in general once implemented